src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,7 +2447,168 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+                Debug.Assert(!nonTypeNonIndexerMembers.Any(s => s.IsAccessor() && ((MethodSymbol)s).AssociatedSymbol.IsIndexer()));
+
+                this.NonTypeNonIndexerMembers = nonTypeNonIndexerMembers;
+                this.StaticInitializers = staticInitializers;
+                this.InstanceInitializers = instanceInitializers;
+                this.IndexerDeclarations = indexerDeclarations;
+                this.StaticInitializersSyntaxLength = staticInitializersSyntaxLength;
+                this.RecordDeclarationWithParameters = recordDeclarationWithParameters;
+                this.InstanceInitializersIndexForRecordDeclarationWithParameters = instanceInitializersIndexForRecordDeclarationWithParameters;
+                this.IsNullableEnabledForInstanceConstructorsAndFields = isNullableEnabledForInstanceConstructorsAndFields;
+                this.IsNullableEnabledForStaticConstructorsAndFields = isNullableEnabledForStaticConstructorsAndFields;
+            }
+        }
+
+        private sealed class MembersAndInitializersBuilder
+        {
+            public ArrayBuilder<Symbol> NonTypeNonIndexerMembers { get; set; }
+            public ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> InstanceInitializers;
+            private bool IsNullableEnabledForInstanceConstructorsAndFields;
+            private bool IsNullableEnabledForStaticConstructorsAndFields;
+
+            public MembersAndInitializersBuilder(DeclaredMembersAndInitializers declaredMembersAndInitializers)
+                : this(
+                    declaredMembersAndInitializers.NonTypeNonIndexerMembers,
+                    declaredMembersAndInitializers.InstanceInitializers,
+                    declaredMembersAndInitializers.IsNullableEnabledForInstanceConstructorsAndFields,
+                    declaredMembersAndInitializers.IsNullableEnabledForStaticConstructorsAndFields)
+            {
+            }
+
+            private MembersAndInitializersBuilder(
+                ImmutableArray<Symbol> nonTypeNonIndexerMembers,
+                ImmutableArray<ImmutableArray<FieldOrPropertyInitializer.Builder>> instanceInitializers,
+                bool isNullableEnabledForInstanceConstructorsAndFields,
+                bool isNullableEnabledForStaticConstructorsAndFields)
+            {
+                Debug.Assert(!nonTypeNonIndexerMembers.IsDefault);
+                Debug.Assert(!instanceInitializers.IsDefault);
+                Debug.Assert(instanceInitializers.All(g => !g.IsDefault));
+
+                this.NonTypeNonIndexerMembers = getAndFillBuilder(nonTypeNonIndexerMembers);
+
+                this.InstanceInitializers = ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>.GetInstance(instanceInitializers.Length);
+                foreach (var group in instanceInitializers)
+                {
+                    this.InstanceInitializers.Add(getAndFillBuilder(group));
```

Comment:
>this.InstanceInitializers.Add(getAndFillBuilder(group)); [](start = 20, length = 56)

Same comment. We should calculate the delta and then combine, if necessary.

Reply:
Then we need another pair of ToReadOnly methods that can handle some immutable arrays and one array builder.
I'm not convinced this is worth the extra complexity/optimization.

---
In reply to: [562784719](https://github.com/dotnet/roslyn/pull/50650#discussion_r562784719) [](ancestors = 562784719)

Reply:
Actually, let me take another stab. We may be able to re-use the merging ToReadOnly for static initializers (avoiding a bunch of extra code).

---
In reply to: [562828740](https://github.com/dotnet/roslyn/pull/50650#discussion_r562828740) [](ancestors = 562828740,562784719)